algorithm/dinic_bidirectional.h

```diff
@@ -0,0 +1,543 @@
+		static constexpr bool log = true;
+		
+		BidirectionalDinic(FlowHypergraph& hg) : BidirectionalDinicBase(hg)
+		{
+			reset();
+		}
+		
+		void reset() {
+		
+		}
+		
+		ScanList& getScanList() {
+			return fqueue;
+		}
+		
+		bool exhaustFlow(CutterState<Type>& cs) {
+			prepare(cs);
+			cs.flowValue += recycleDatastructuresFromGrowReachablePhase(cs);
+			bool hasCut = false;
+			while (cs.flowValue <= upperFlowBound) {
+				hasCut = !buildLayeredNetwork(cs, true, false);
+				LOGGER << V(hasCut);
+				if (hasCut || cs.flowValue >= upperFlowBound) {
+					break;
+				}
+				else {
+					cs.flowValue += augmentFlowInLayeredNetwork(cs);
+				}
+			}
+			finish(cs);
+			LOGGER << V(cs.flowValue);
+			return hasCut;
+		}
+		
+		Flow growFlowOrSourceReachable(CutterState<Type>& cs) {
+			prepare(cs);
+			Flow f = 0;
+			if (buildLayeredNetwork(cs, true, false))
+				f += augmentFlowInLayeredNetwork(cs);
+			finish(cs);
+			return f;
+		}
+		
+		void growReachable(CutterState<Type>& cs) {
+			prepare(cs);
+			bool found_target = buildLayeredNetwork(cs, false, true);
+			assert(!found_target); unused(found_target);
+			finish(cs);
+		}
+		
+		Flow recycleDatastructuresFromGrowReachablePhase(CutterState<Type> &cs) {
+			/*
+			if (!cs.augmentingPathAvailableFromPiercing || std::none_of(cs.sourcePiercingNodes.begin(), cs.sourcePiercingNodes.end(),
+																		[](const auto& sp) { return sp.isReachableFromOppositeSide; })) {
+				return 0;
+			}
+			cs.flipViewDirection();
+			resetSourcePiercingNodeDistances(cs, false);
+			Flow f = augmentFlowInLayeredNetwork(cs);
+			resetSourcePiercingNodeDistances(cs);
+			cs.flipViewDirection();
+			return f;
+			 */
+			return 0;		// ignore for now
+		}
+		
+	private:
+		
+		void prepare(CutterState<Type>& cs) {
+			previous_cutter_state_direction = cs.currentViewDirection();
+			if (previous_cutter_state_direction != 0) {
+				cs.flipViewDirection();
+			}
+			assert(cs.currentViewDirection() == 0);
+		}
+		
+		void finish(CutterState<Type>& cs) {
+			for (auto& sp : cs.sourcePiercingNodes) {
+				cs.n.distance[sp.node] = cs.n.sourceSettledDistance;
+			}
+			for (auto& tp : cs.targetPiercingNodes) {
+				cs.n.distance[tp.node] = cs.n.targetSettledDistance;
+			}
+			if (previous_cutter_state_direction != 0) {
+				cs.flipViewDirection();
+			}
+			assert(cs.currentViewDirection() == previous_cutter_state_direction);
+		}
+		
+		
+		enum class STATE {
+			NONE, FLOW_SENDING, ALL,
+		};
+		std::vector<STATE> scan_for_backward;
+		
+		bool buildLayeredNetwork(CutterState<Type>& cs, const bool augment_flow, const bool only_forward) {
+			unused(augment_flow);	//for debug builds only
+			//cs.clearForSearch();
+			assert(cs.currentViewDirection() == 0);
+			
+			if (scan_for_backward.size() != hg.numHyperedges()) {
+				scan_for_backward.resize(hg.numHyperedges(), STATE::NONE);
+			}
+			std::fill(scan_for_backward.begin(), scan_for_backward.end(), STATE::NONE);
```

Comment:
this should be lazy

Reply:
Yup. I'm still looking for a way to get rid of this array altogether